File: test/main/java/org/apache/prepbuddy/utils/PivotTableTest.java
package org.apache.prepbuddy.utils;

import org.apache.prepbuddy.SparkTestCase;
import org.apache.prepbuddy.datacleansers.imputation.Probability;
import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.prepbuddy.transformers.TransformationFunction;
import org.apache.spark.api.java.JavaRDD;
import org.junit.Test;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;

public class PivotTableTest extends SparkTestCase {
    @Test
    public void shouldGiveDefaultValueWhenColumnKeyIsNotPresent() {
        PivotTable<Integer> pivotTable = new PivotTable<>(0);
        pivotTable.addEntry("firstRow", "firstColumn", 5);

        int value = pivotTable.valueAt("firstRow", "secondColumn");
        assertEquals(0, value);
    }

    @Test
    public void shouldTransformThePivotTable() {
        JavaRDD<String> initialDataSet = javaSparkContext.parallelize(Arrays.asList(
                "known,new,long,home,skips",
                "unknown,new,short,work,reads",
                "unknown,follow Up,long,work,skips",
                "known,follow Up,long,home,skips",
                "known,new,short,home,reads",
                "known,follow Up,long,work,skips",
                "unknown,follow Up,short,work,skips",
                "unknown,new,short,work,reads",
                "known,follow Up,long,home,skips",
                "known,new,long,work,skips",
                "unknown,follow Up,short,home,skips",
                "known,new,long,work,skips",
                "known,follow Up,short,home,reads",
                "known,new,short,work,reads",
                "known,new,short,home,reads",
                "known,follow Up,short,work,reads",
                "known,new,short,home,reads",
                "unknown,new,short,work,reads"
        ));
        TransformableRDD initialRDD = new TransformableRDD(initialDataSet);
        final long count = initialRDD.count();
        PivotTable<Integer> pivotTable = initialRDD.pivotByCount(4, new int[]{0, 1, 2, 3});
        PivotTable<Probability> transform = (PivotTable<Probability>) pivotTable.transform(new TransformationFunction<Integer, Probability>() {
            @Override
            public Probability transform(Integer integer) {
                return new Probability((double) integer.intValue() / count);
            }

            @Override
            public Probability defaultValue() {
                return new Probability(0);
            }
        });
        Probability expected = new Probability((double) 7 / 18);
        Probability actual = transform.valueAt("skips", "long");

        assertEquals(expected, actual);

        Probability expectedZero = new Probability(0);
        assertEquals(expectedZero, transform.valueAt("reads", "long"));
    }
}

File: src/test/scala/org/apache/datacommons/prepbuddy/utils/PivotTableTest.scala
package org.apache.datacommons.prepbuddy.utils

import org.apache.datacommons.prepbuddy.SparkTestCase
import org.apache.datacommons.prepbuddy.rdds.TransformableRDD
import org.apache.spark.rdd.RDD

class PivotTableTest extends SparkTestCase {
    test("pivotByCount should give pivoted table of given row and column indexes") {
        val pivotTable: PivotTable[Int] = new PivotTable[Int](0)
        pivotTable.addEntry("row", "column1", 5)

        val expected: Int = 5
        val actual: Int = pivotTable.valueAt("row", "column1")
        assert(expected == actual)

        assert(0 == pivotTable.valueAt("row", "column"))

    }

    test("pivotTable can be transformed") {
        val pivotTable: PivotTable[Int] = new PivotTable[Int](0)
        pivotTable.addEntry("row", "column1", 5)
        pivotTable.addEntry("row", "column2", 6)


        val transformed: PivotTable[Double] = pivotTable.transform((value: Any) => {
            0.9
        }, 10.0).asInstanceOf[PivotTable[Double]]


        val value: Any = transformed.valueAt("row", "column1")
        assert(0.9 == value)

        assert(10.0 == transformed.valueAt("row", "column"))
    }

    test("pivotByCount should give pivoted table of given column") {
        val dataSet = Array("known,new,long,home,skips",
            "unknown,new,short,work,reads",
            "unknown,follow Up,long,work,skips",
            "known,follow Up,long,home,skips",
            "known,new,short,home,reads",
            "known,follow Up,long,work,skips",
            "unknown,follow Up,short,work,skips",
            "unknown,new,short,work,reads",
            "known,follow Up,long,home,skips",
            "known,new,long,work,skips",
            "unknown,follow Up,short,home,skips",
            "known,new,long,work,skips",
            "known,follow Up,short,home,reads",
            "known,new,short,work,reads",
            "known,new,short,home,reads",
            "known,follow Up,short,work,reads",
            "known,new,short,home,reads",
            "unknown,new,short,work,reads")
        val initialDataSet: RDD[String] = sparkContext.parallelize(dataSet)
        val initialRDD: TransformableRDD = new TransformableRDD(initialDataSet)
        val pivotTable: PivotTable[Integer] = initialRDD.pivotByCount(4, Seq[Int](0, 1, 2, 3))

        val expected: Int = 7
        val actual: Integer = pivotTable.valueAt("skips", "long")
        assert(expected == actual)

        assert(0 == pivotTable.valueAt("reads", "long"))
    }
}
